File: config-common/src/main/java/io/github/solusmods/eternalcore/config/impl/network/s2c/SyncConfigToClientPayload.java
package io.github.solusmods.eternalcore.config.impl.network.s2c;


import dev.architectury.networking.NetworkManager;
import dev.architectury.utils.Env;
import io.github.solusmods.eternalcore.config.ConfigRegistry;
import io.github.solusmods.eternalcore.config.EternalCoreConfig;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import org.jetbrains.annotations.NotNull;

import java.util.Map;

/**
 * Handles the syncing of configuration data from the server to the client.
 * This is triggered when a player joins a multiplayer server.
 */
public record SyncConfigToClientPayload(
        Map<String, String> configData
) implements CustomPacketPayload {
    public static final Type<SyncConfigToClientPayload> TYPE = new Type<>(EternalCoreConfig.create("sync_config_to_client"));
    public static final StreamCodec<FriendlyByteBuf, SyncConfigToClientPayload> STREAM_CODEC = CustomPacketPayload.codec(SyncConfigToClientPayload::encode, SyncConfigToClientPayload::new);

    public SyncConfigToClientPayload(FriendlyByteBuf buf) {
        this(buf.readMap(FriendlyByteBuf::readUtf, FriendlyByteBuf::readUtf));
    }

    public void encode(FriendlyByteBuf buf) {
        buf.writeMap(this.configData, FriendlyByteBuf::writeUtf, FriendlyByteBuf::writeUtf);
    }

    public void handle(NetworkManager.PacketContext context) {
        if (context.getEnvironment() != Env.CLIENT) return;
        context.queue(() -> ConfigRegistry.loadConfigSyncData(this.configData));
    }

    @NotNull
    public CustomPacketPayload.Type<SyncConfigToClientPayload> type() {
        return TYPE;
    }
}

File: config-common/src/main/kotlin/io/github/solusmods/eternalcore/config/impl/network/s2c/SyncConfigToClientPayload.kt
package io.github.solusmods.eternalcore.config.impl.network.s2c

import dev.architectury.networking.NetworkManager
import dev.architectury.utils.Env
import io.github.solusmods.eternalcore.config.ConfigRegistry.Companion.loadConfigSyncData
import io.github.solusmods.eternalcore.config.EternalCoreConfig.create
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.codec.StreamDecoder
import net.minecraft.network.codec.StreamEncoder
import net.minecraft.network.protocol.common.custom.CustomPacketPayload


/**
 * Handles the syncing of configuration data from the server to the client.
 * This is triggered when a player joins a multiplayer server.
 */
@JvmRecord
data class SyncConfigToClientPayload(
    val configData: MutableMap<String?, String?>?
) : CustomPacketPayload {
    constructor(buf: FriendlyByteBuf) : this(
        buf.readMap<String?, String?>(
            StreamDecoder { obj: FriendlyByteBuf? -> obj!!.readUtf() },
            StreamDecoder { obj: FriendlyByteBuf? -> obj!!.readUtf() })
    )

    fun encode(buf: FriendlyByteBuf) {
        buf.writeMap<String, String>(
            this.configData!!,
            StreamEncoder { obj: FriendlyByteBuf?, string: String? -> obj!!.writeUtf(string) },
            StreamEncoder { obj: FriendlyByteBuf?, string: String? -> obj!!.writeUtf(string) })
    }

    fun handle(context: NetworkManager.PacketContext) {
        if (context.environment != Env.CLIENT) return
        context.queue { loadConfigSyncData(this.configData) }
    }

    override fun type(): CustomPacketPayload.Type<SyncConfigToClientPayload> {
        return TYPE
    }

    companion object {
        val TYPE: CustomPacketPayload.Type<SyncConfigToClientPayload> =
            CustomPacketPayload.Type<SyncConfigToClientPayload>(
                create("sync_config_to_client")
            )
        val STREAM_CODEC: StreamCodec<FriendlyByteBuf, SyncConfigToClientPayload> =
            CustomPacketPayload.codec<FriendlyByteBuf, SyncConfigToClientPayload>(
                { obj: SyncConfigToClientPayload?, buf: FriendlyByteBuf? -> obj!!.encode(buf!!) },
                { buf: FriendlyByteBuf? ->
                    SyncConfigToClientPayload(
                        buf!!
                    )
                })
    }
}
